Adding explanation for editing permissions 20230825 (#6606)
opensearch-project/documentation-website: _security/access-control/document-level-security.md

# Document-level security (DLS)

Document-level security lets you restrict a role to a subset of documents in an index.
For more information about OpenSearch users and roles, see the [documentation](https://opensearch.org/docs/latest/security/access-control/users-roles/#create-roles).

Use the following steps to get started with document-level and field-level security:
1. Open OpenSearch Dashboards.
2. Choose **Security** > **Roles**.
3. Select **Create Role** and provide a name for the role.
4. Review the **Index permissions** section and any necessary [index permissions](https://opensearch.org/docs/latest/security/access-control/permissions/) for the role. 

> 5. Add document-level security, with the addition of a Domain Specific Language (DSL) query inside the `Document level security - optional` section. A typical request sent to the `_search` API includes `{ "query": { ... } }` around the query, but with document-level security in OpenSearch Dashboards, you only need to specify the query itself. For example, the following DSL query specifies that for the new role to have access to a document, the query's `genres` field must include `Comedy`:

Reviewer: ```suggestion
5. Add document-level security, with the addition of a domain-specific language (DSL) query in the `Document level security - optional` section. A typical request sent to the `_search` API includes `{ "query": { ... } }` around the query, but with document-level security in OpenSearch Dashboards, you only need to specify the query itself. For example, the following DSL query specifies that for the new role to have access to a document, the query's `genres` field must include `Comedy`:
```

# Document-level security (DLS)

Document-level security lets you restrict a role to a subset of documents in an index.

> For more information about users and roles in OpenSearch[, see the documentation.](https://opensearch.org/docs/latest/security/access-control/users-roles/#create-roles)

Reviewer 2: ```suggestion
For more information about users and roles in OpenSearch, see the [documentation](https://opensearch.org/docs/latest/security/access-control/users-roles/#create-roles).
```

> For more information about users and roles in OpenSearch, see the [documentation](https://opensearch.org/docs/latest/security/access-control/users-roles/#create-roles).

Reviewer: ```suggestion
For more information about OpenSearch users and roles, see the [documentation](https://opensearch.org/docs/latest/security/access-control/users-roles/#create-roles).
```

> The easiest way to get started with document- and field-level security is:

Reviewer 2: ```suggestion
Use the following steps to get started with document-level and field-level security:
```

> 1. Open OpenSearch Dashboards 

Reviewer 2: ```suggestion
1. Open OpenSearch Dashboards.
```

> 2. Choose **Security**

Reviewer 2: ```suggestion
2. Choose **Security** > **Roles**
```

> 3. Choose **Roles** 

Reviewer 2: ```suggestion

```

> 4. Create a new role

Reviewer 2: ```suggestion
3. Select **Create Role** and provide a name for the role.
```

> 5. Review the **Index permissions** section

Reviewer 2: What are users looking for as they review the **Index permissions** section?

Author: Inside the 'index permissions' area/heading, is the DSL multiline text box, so it was just a step to get to the DSL area.

> 5. Review the **Index permissions** section

Reviewer 2: ```suggestion
4. Review the **Index permissions** section and any necessary [index permissions](https://opensearch.org/docs/latest/security/access-control/permissions/) for the role. 
```

> 6. Add document-level security with the addition of a DSL query to the `Document level security - optional` section

Reviewer 2: ```suggestion
5. Add document-level security with the addition of a DSL query inside the `Document level security - optional` section.  A typical request to the `_search` API includes the `{ "query": { ... } }` around the query, but when Document-level security in OpenSearch Dashboards, you only need to specify the query itself. For example, the following DSL query specifies that for the new role to have access to a document, the query's `genres` field must include `Comedy`:
```

Then add the code example underneath.

> 5. Add document-level security with the addition of a DSL query inside the `Document level security - optional` section.  A typical request to the `_search` API includes the `{ "query": { ... } }` around the query, but when Document-level security in OpenSearch Dashboards, you only need to specify the query itself. For example, the following DSL query specifies that for the new role to have access to a document, the query's `genres` field must include `Comedy`:

Reviewer: ```suggestion
5. Add document-level security, with the addition of a DSL query inside the `Document level security - optional` section. A typical request sent to the `_search` API includes `{ "query": { ... } }` around the query, but with document-level security in OpenSearch Dashboards, you only need to specify the query itself. For example, the following DSL query specifies that for the new role to have access to a document, the query's `genres` field must include `Comedy`:
```

   - For example, the following DSL could be added in the `Document level security - optional` text box.
   - This query specifies that for the role to have access to a document, its `genres` field must include `Comedy`
   - A typical request to the `_search` API includes `{ "query": { ... } }` around the query, but in this case, you only need to specify the query itself.
   ```json
   {
     "bool": {
       "must": {
         "match": {
           "genres": "Comedy"
         }
       }
     }
   }
   ```

>    - ![Document- and field-level security screen in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/images/security-dls.png)

Reviewer 2: Add a space between this line and the code example.

```suggestion

   - ![Document- and field-level security screen in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/images/security-dls.png)
```

>    - A typical request to the `_search` API includes `{ "query": { ... } }` around the query, but in this case, you only need to specify the query itself.

Reviewer: I'm not following this bulleted list. It appears to contain content duplicated from above, and the structure doesn't work in terms of introducing the query.

>    - A typical request to the `_search` API includes `{ "query": { ... } }` around the query, but in this case, you only need to specify the query itself.

Reviewer: ```suggestion
   - A typical request to the `_search` API includes `{ "query": { ... } }` around the query, but, in this case, you only need to specify the query itself.
```

Reviewer 2: We can likely just remove this bulleted list.

> ## Updating roles by accessing the Rest API 

Reviewer: ```suggestion
## Updating roles by accessing the REST API 
```

When a user has both Role A and Role B permissions, the query results are filtered based on Role A's DLS, even though Role B doesn't use DLS. The DLS settings are retained, and the returned access is appropriately restricted. 

When a user is assigned both Role A and Role B and the `plugins.security.dfm_empty_overrides_all` setting is enabled, Role B's permissions Role B's permissions will override Role A's restrictions, allowing that user to access all documents. This ensures that the role without DLS takes precedence in the search query response.

> ### Editing Document-Level Permissions

Reviewer 2: These need to be in sentence case to fix the build error.

```suggestion
### Editing document-level permissions
```

> To adjust document-level permissions in OpenSearch, you can tailor existing roles or create new ones with customized configurations.

Reviewer 2: ```suggestion
To adjust document-level permissions in OpenSearch, you can add the permissions to existing roles or create new ones with customized configurations.
```

> To modifying an existing role, open OpenSearch Dashboards and choose Security and then choose Roles. Within the list of available roles locate the role to be changed, click on the role to inspect its current permissions and configurations.

Reviewer 2: ```suggestion
To modifying an existing role: 
1. Open OpenSearch Dashboards. 
2. Select  **Security > Roles**. This page lists all the available roles within your configuration.  
3. Locate the role to you want to change and select it to inspect its current permissions and configurations.
```

> Within the role details, you can change index permissions in the specified index permissions section. Here, you can update index patterns, document-level security queries (DLS), and allowed actions associated with the role.

Reviewer 2: ```suggestion
4. Within the role details, you can change index permissions in the specified index permissions section, such as can index patterns, document-level security queries (DLS), and allowed actions associated with the role. The following example adds DLS permissions and lists the actions allowed for the role:
```

> Example of DLS Configuration in roles.yml:

Reviewer 2: ```suggestion

```

> ```

Reviewer 2: Do we have a screenshot of where they would add this in Dashboards or can this only be changed in `roles.yml`?

Reviewer 2: Also, we'll need to adjust the space on the code example so it fits under step 4.

{
  "index_permissions": [
    {
      "index_patterns": [
        "example-index"
      ],
      "dls": "[.. some DLS here ..]",
      "allowed_actions": [
        "indices:data/read/search",
      ]
    }
  ]
}
```

> Edit and save the configuration files, to upload the changes into the security index execute `securityadmin.sh`

Reviewer 2: I'm slightly confused on where this action would take place? On one hand, we tell people they can open roles in Dashboards and modify them there. However, `securityadmin.sh` would be run in the cluster's console or host? Are both paths viable? Can roles be edited in Dashboards?

Author: Hi @Naarcha-AWS , on a further look after your comment I realised I could be clearer in my approach to updating the documentation.
I have rewritten this section to be in the roles and users area, and I have also updated the DLS documentation page. Please let me know your thoughts - thanks a million
